### PR TITLE
status check enforcement

### PR DESCRIPTION
**Pull request by ajkr**
Tried making Status object enforce that it is checked in some way. In cases it is not checked, `PermitUncheckedError()` must be called explicitly.

Added a way to run tests (`ASSERT_STATUS_CHECKED=1 make -j48 check`) on a
whitelist. The effort appears significant to get each test to pass with
this assertion, so I only fixed up enough to get one test (`options_test`)
working and added it to the whitelist.

### CODE REVIEW

File: include/rocksdb/io_status.h [hunk 1]
```diff
@@ -204,14 +211,18 @@ inline IOStatus& IOStatus::operator=(IOStatus&& s)
...
+#endif  // ROCKSDB_ASSERT_STATUS_CHECKED
     code_ = std::move(s.code_);
     s.code_ = kOk;
     subcode_ = std::move(s.subcode_);
     s.subcode_ = kNone;
     retryable_ = s.retryable_;
     data_loss_ = s.data_loss_;
     scope_ = s.scope_;
-    scope_ = kIOErrorScopeFileSystem;
+    s.scope_ = kIOErrorScopeFileSystem;
```
Comment (pdillinger) [inline include/rocksdb/io_status.h:225]: Nice catch

File: include/rocksdb/io_status.h [hunk 2]
```diff
@@ -204,14 +211,18 @@ inline IOStatus& IOStatus::operator=(IOStatus&& s)
 #endif
 {
   if (this != &s) {
+#ifdef ROCKSDB_ASSERT_STATUS_CHECKED
+    s.checked_ = true;
```
Comment (pdillinger) [inline include/rocksdb/io_status.h:215]: Why does Status ctor set to false but set to true after move?

Comment (pdillinger) [reply include/rocksdb/io_status.h:215]: OK, that's obvious. Nevermind.

**Comment by ajkr** [comment]
Thanks for the review!

**Pull request closed without merging**